File: Python/数据结构/Treearray.py
class Tree:
    def __init__(self, a, init_val: int = 0) -> None:
        if isinstance(a, int):
            n = a
            self.tree = [0] * (n + 1)
            for i in range(1, n + 1):
                self.tree[i] += init_val
                j = i + (i & -i)
                if j <= n:
                    self.tree[j] += self.tree[i]
        else:
            n = len(a)
            self.tree = [0] * (n + 1)
            for i, x in enumerate(a, 1):
                self.tree[i] += x
                j = i + (i & -i)
                if j <= n:
                    self.tree[j] += self.tree[i]

    def add(self, i: int, val: int = 1) -> None:
        i += 1
        while i < len(self.tree):
            self.tree[i] += val
            i += i & -i

    def pre(self, i: int) -> int:
        res = 0
        i += 1
        while i > 0:
            res += self.tree[i]
            i &= i - 1
        return res

    def query(self, l: int, r: int) -> int:
        if r < l:
            return 0
        return self.pre(r) - self.pre(l - 1)

    def lower_bound(self, val: int) -> int:
        n = len(self.tree) - 1
        res = 0
        cur = 0
        bit = 1 << (n.bit_length() - 1) if n else 0
        while bit:
            nxt = res + bit
            if nxt <= n and cur + self.tree[nxt] < val:
                res = nxt
                cur += self.tree[nxt]
            bit >>= 1
        return res


class MaxTree:
    def __init__(self, a, identity: float = float("-inf")) -> None:
        self.identity = identity
        if isinstance(a, int):
            self.tree = [identity] * (a + 1)
            return
        n = len(a)
        self.tree = [identity] * (n + 1)
        for i, x in enumerate(a, 1):
            self.tree[i] = max(self.tree[i], x)
            j = i + (i & -i)
            if j <= n:
                self.tree[j] = max(self.tree[j], self.tree[i])

    def chmax(self, i: int, val: int) -> None:
        i += 1
        while i < len(self.tree):
            self.tree[i] = max(self.tree[i], val)
            i += i & -i

    def pre(self, i: int) -> float:
        res = self.identity
        i += 1
        while i > 0:
            res = max(res, self.tree[i])
            i &= i - 1
        return res


# 用法（下标均为 0-based；区间均为闭区间）
# tree = Tree(a)
# tree = Tree(n, init_val)
# tree.add(i, x)
# s = tree.pre(i)          # [0, i] 的和
# s = tree.query(l, r)     # [l, r] 的和
# i = tree.lower_bound(x)  # 最小的 i 使 pre(i) >= x；不存在返回 n
# lower_bound 要求前缀和单调，通常即数组元素非负
#
# tree = MaxTree(a)
# tree.chmax(i, x)         # 只支持单点变大
# mx = tree.pre(i)         # [0, i] 的最大值

File: C++/数据结构/Treearray.cpp
template <typename T = long long>
class Tree {
    vector<T> tree;

public:
    // 构造函数：初始化大小为 n 的树状数组，初始所有元素值为 val（外部表现为 0-based）
    Tree(int n, T val = 0) : tree(n + 1) {
        for (int i = 1; i <= n; i++) {
            tree[i] += val;
            int nxt = i + (i & -i);
            if (nxt <= n) {
                tree[nxt] += tree[i];
            }
        }
    }

    // 构造函数：使用给定的 vector 在 O(N) 时间内快速初始化建树
    Tree(const vector<T>& data) {
        int n = data.size();
        tree.resize(n + 1);
        for (int i = 1; i <= n; i++) {
            tree[i] += data[i - 1];  // data是 0-based
            int nxt = i + (i & -i);
            if (nxt <= n) {
                tree[nxt] += tree[i];
            }
        }
    }

    // 单点修改：将 0-based 下标 i 处的元素增加 val
    void add(int i, T val = 1) {
        for (++i; i < tree.size(); i += i & (-i)) {
            tree[i] += val;
        }
    }

    // 前缀求和：计算 0-based 下标区间 [0, i] 内的所有元素之和
    T pre(int i) const {
        T res = 0;
        for (++i; i > 0; i &= i - 1) {
            res += tree[i];
        }
        return res;
    }

    // 区间求和：计算 0-based 下标区间 [l, r] 内的所有元素之和
    T query(int l, int r) const {
        if (r < l) {
            return 0;
        }
        return pre(r) - pre(l - 1);  // 当 l=0 时, pre(-1) 会合理地返回 0
    }

    // 树上二分查找：返回满足前缀和 >= val 的最小 0-based 下标
    int lower_bound(T val) const {
        int w = bit_width(tree.size() - 1);
        int res = 0;
        T s = 0;
        for (int i = w - 1; i >= 0; i--) {
            int nxt = res + (1 << i);
            if (nxt < tree.size() && tree[nxt] + s < val) {
                res += (1 << i);
                s += tree[nxt];
            }
        }
        return res;  // 返回 0-based 下标：内部 1-based 下标为 res + 1，因此 0-based 为 res
    }
};

template <typename T = long long>
class MaxTree {
    vector<T> tree;
    T identity;

public:
    // 构造函数：初始化大小为 n 的树状数组（外部表现为 0-based）
    explicit MaxTree(int n, T identity = numeric_limits<T>::lowest())
        : tree(n + 1, identity), identity(identity) {}

    // 构造函数：使用给定的 vector 在 O(N) 时间内快速初始化建树
    explicit MaxTree(const vector<T>& data, T identity = numeric_limits<T>::lowest())
        : tree(data.size() + 1, identity), identity(identity) {
        int n = data.size();
        for (int i = 1; i <= n; i++) {
            tree[i] = max(tree[i], data[i - 1]);  // data 是 0-based
            int nxt = i + (i & -i);
            if (nxt <= n) {
                tree[nxt] = max(tree[nxt], tree[i]);
            }
        }
    }

    // 单点修改：将 0-based 下标 i 处的值与 val 取最大值（只支持值变大）
    void chmax(int i, T val) {
        for (++i; i < tree.size(); i += i & -i) {
            tree[i] = max(tree[i], val);
        }
    }

    // 前缀最大值：计算 0-based 下标区间 [0, i] 内的最大值
    T pre(int i) const {
        T res = identity;
        for (++i; i > 0; i &= i - 1) {
            res = max(res, tree[i]);
        }
        return res;
    }
};
